rfcs/2020-09-28-cloudwatch-metrics.md

```diff
@@ -0,0 +1,181 @@
+before looking at this if you you are unfamiliar with CloudWatch.
+
+## Scope
+
+This RFC will simply cover a new `aws_cloudwatch_metrics` source.
+
+## Motivation
+
+Users want to collect and forward metrics from AWS CloudWatch to monitor
+infrastructure and deployed services in AWS.
+
+## Internal Proposal
+
+We will add a `aws_cloudwatch_metrics` source that scrapes the AWS API
+([`GetMetricData`](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_GetMetricData.html))
+on a user-defined interval to collect metrics and forward them through the
+user's pipeline.
+
+## Doc-level Proposal
+
+The configuration of the new `aws_cloudwatch_metrics` source will look like:
+
+```toml
+[sources.cloudwatch]
+  type = "aws_cloudwatch_metrics"
+  assume_role = "arn:aws:iam::123456789098:role/my_role" # optional, no default
+  endpoints = ["127.0.0.0:5000/path/to/service"] # optional, no default, relevant when regions = []
+  regions = ["us-east-1"] # required, required when endpoints unspecified; no default
+
+  period_secs = 300 # period (s) to aggregate metrics over, optional, can be overridden at metric level, default 300
+  delay_secs = 300 # delay collection by value (s), used to avoid collecting data that has not fully been processed by CloudWatch, optional, default 300
+  interval_secs = 300 # interval to scrape metrics; should be a multiple of "period"; default 300
+
+  metrics_refresh_interval_secs = 900 # interval to refresh available metrics for queried namespaces if globbing or all metrics are used; default 900
+
+  # Request
+  request.in_flight_limit = 25 # optional, default, requests
+  request.rate_limit_duration_secs = 1 # optional, default, seconds
+  request.rate_limit_num = 25 # optional, default
+  request.retry_attempts = 18446744073709551615 # optional, default
+  request.retry_initial_backoff_secs = 1 # optional, default, seconds
+  request.retry_max_duration_secs = 10 # optional, default, seconds
+  request.timeout_secs = 30 # optional, default, seconds
+
+  [[sources.cloudwatch.metrics]]
+    namespace = "AWS/EC2" # optional; supports globbing
+    names = ["EBSReadOps", "EBSReadBytes", "Network*"] # optional; defaults to all metrics in namespace, ["*"], (refreshed on interval); supports globbing
+    dimensions.InstanceId = "i-05517fbc2e6124dfb" # optional; supported dimensions differ by namespace and metric; supports globbing
+    statistics = [ "average", "sum", "minimum", "maximum", "sample_count" ] # statistics to collect; can also contain extended statistics like p99; default: [ "average", "sum", "minimum", "maximum", "sample_count" ]
+    period_secs = 300 # period (s) to aggregate metrics over; defaults to top-level `period` setting; top-level interval should be a multiple of this and any other defined periods
+```
+
+We could alternatively model dimensions as another table:
+
+```toml
+[[sources.cloudwatch.metrics.dimensions]]
+  key = "InstanceId"
+  value = "i-05517fbc2e6124dfb"
+```
```

Comment:
I like this approach.

It seems like the single-dimension shorthand isn't that more verbose either:

```toml
dimensions.InstanceId = "i-05517fbc2e6124dfb" 							# before
dimensions = [{ key = "InstanceId", value = "i-05517fbc2e6124dfb" }]	# after
```

Reply:
It looks like we have some precedent for representing key/value config as a map in TOML:

https://vector.dev/docs/reference/transforms/add_tags/ (tags)
https://vector.dev/docs/reference/sinks/http/ (headers)

Given that, I'll stick with that for this RFC, but we can always adjust if you feel strongly about the table format.